### ISSUE TITLE
Return value from serialisation

### ISSUE THREAD

**Issue by angusturner** [patrick-kidger/equinox#187]
Hi,

Another feature request / suggestion! I think it would be cool if there was an in-built way to get a pickle-able / serializable copy of the model state. For example, `equinox.tree_serialise_leaves` automatically saves state to a file, but it would be great if there was a way to return the state instead. 

My reasoning is, it is often useful to save a bunch of extra state with an experiment, aside from the model weights themselves (e.g. the epoch number, learning rate schedule etc). 

e.g) ideally I would like to do:

```
model_state = eqx.tree_get_serialisable(model)
torch.save({
      'model': model_state,
      'epoch': ...
})
```

(Or, replace `torch.save` with pickle or whatever. I just happen to be in a mixed torch/jax code-base atm).

Let me know your thoughts!

**Comment by paganpasta** [comment]
Hi,

You can try and wrap the `(key, values)` in a dict and {de}serialise this dict (as a new PyTree). I have not tried it with the `opt_state` from the optimiser, but I think it should work.

**Comment by patrick-kidger** [comment]
To expand on what @paganpasta is saying.

In a typical pure-JAX setup, you should be able to just directly serialise the model, opt state etc. all in one go:
```python
eqx.tree_serialise_leaves("out.eqx", (model, opt_state, ...))
```
This works because Equinox models are just PyTrees, and so `(model, opt_state)` is also just a PyTree, and serialising PyTrees is exactly the job of `eqx.tree_serialise_leaves`. (It doesn't have to just be a model.)

Now admittedly I've not thought about how any of this might work ina mixed PyTorch/JAX codebase.

A function which returns the serialisable state will depend on what you count as serialisable. It could just be the identity function `lambda x: x` if you're happy pickling the whole thing.

If you just want the JAX arrays, for example (and are happy to promise that you'll reconstitute anything else yourself when you deserialise) you could do something like `list(filter(eqx.is_array, jax.tree_leaves(model)))`.

Possibly you might be able to still use `eqx.tree_serialise_leaves` if your PyTorch component is also a PyTree.

Do any of the above approaches work for you?

**Comment by angusturner** [comment]
Thanks, that's quite helpful. Based on your comments and some of the earlier discussion around serialization, I will do something like:

```
def get_model_state(model, filter_spec=eqx.is_array):
    tree_weights, tree_other = eqx.partition(model, filter_spec)
    flattened_weights, _ = jtu.tree_flatten(tree_weights)
    return flattened_weights
    
def set_model_state(model, weights, filter_spec=eqx.is_array):
    tree_weights, tree_other = eqx.partition(model, filter_spec)
    tree_def = jtu.tree_structure(tree_weights)
    loaded_weights = jtu.tree_unflatten(tree_def, weights)
    return eqx.combine(loaded_weights, tree_other)
```

I do not actually need to mix/match PyTorch/Jax in a single model, but am just experimenting with adding Eqx support to my current framework without breaking the API too much! I think this should to the trick. 

Not sure if such a feature (i.e. returning state instead of saving) would be valuable in the base library, but happy to close this issue if not!

**Comment by patrick-kidger** [comment]
LGTM!

Yup, right now I don't think we  need to add such a function.
